astroid/nodes/scoped_nodes/scoped_nodes.py

```diff
@@ -1583,17 +1585,13 @@ def infer_call_result(
                 )
             if isinstance(metaclass, ClassDef):
                 try:
-                    class_bases = [
-                        # Find the first non-None inferred base value
-                        next(
-                            b
-                            for b in arg.infer(
-                                context=context.clone() if context else context
-                            )
-                            if not (isinstance(b, Const) and b.value is None)
-                        )
-                        for arg in caller.args[1:]
-                    ]
+                    # Find the first non-None inferred base value
+                    get_base = lambda arg: next(  # pylint: disable=unnecessary-lambda-assignment # noqa
```

Comment:
Hi, sorry for the drive-by review. Just wondering: can we do this without the lambda & disable?

Reply:
Absolutely, but the formatter then turns it into a monstrosity with several single word lines. This one looks clearer. However, it is subjective and I will have no problems inlining the lambda.

Reply:
What about without a lambda at all?

Reply:
> What about without a lambda at all?

Oh, yes, that's what I meant. I'm comparing the situation without a lambda = lambda inlined (weird formatting) vs with a lambda (the present).

Reply:
Sorry, I guess I was wondering about extracting a function? That's the recommended prescription for this lint message.

Reply:
> Sorry, I guess I was wondering about extracting a function? That's the recommended prescription for this lint message.

Done!

Reply:
AFAICS, the underlying issue that necessitates that function is the `six.with_metaclass` issue (#5935). It's caused by the fact that `Enum` is defined two times in the `enum` module: first as `None`, second as the actual class.

So, the problem is that symbols can be defined several times. This workaround with skipping `None` values is unsatisfactory, because it one hand behaves too generally for handling just the `Enum` issue, but not generally enough to handle the wider problem of redefinition of symbols.

Perhaps, if we wanted to solve the specific Enum problem, we could do it with an `Enum` transform.
However, if we wanted to solve the larger problem, there is one heuristic that seems to make sense to me. When you import a symbol, only the last redefinition of that symbol is imported. Thus, we shouldn't be looking for the first «non-None», we should be looking for the _last_ definition (specifically, when we are inferring imports). 

These are just thoughts, but do tell me if you think there is some worth in it.

Reply:
Absolutely. Elsewhere for functions we (as of recently) take the last definition unless it's a property and decorated with setter. On mobile so don't have a link. But you're right on.

Reply:
Got it. Based on our conversation, I've changed the the `infer_first_not_none` to `infer_last`, although it's better be done in the inference function for imports. That would be too much out of scope of this commit though.